src/main/groovy/com/github/winplay02/gitcraft/pipeline/Pipeline.java

```diff
@@ -0,0 +1,96 @@
+import com.github.winplay02.gitcraft.util.RepoWrapper;
+
+public class Pipeline {
+
+	private final Map<StepResult, Path> resultFiles;
+	private final Map<MinecraftJar, Path> minecraftJars;
+
+	public Pipeline() {
+		this.resultFiles = new HashMap<>();
+		this.minecraftJars = new EnumMap<>(MinecraftJar.class);
+	}
+
+	Path initResultFile(Step step, StepWorker.Context context, StepResult resultFile) {
+		Path path = step.getResultFile(resultFile, context);
+		resultFiles.put(resultFile, path);
+		return path;
+	}
+
+	public Path getResultFile(StepResult resultFile) {
+		return resultFiles.get(resultFile);
+	}
+
+	public Path getMinecraftJar(MinecraftJar minecraftJar) {
+		return minecraftJars.get(minecraftJar);
+	}
+
+	public void run(RepoWrapper repository, MinecraftVersionGraph versionGraph, OrderedVersion minecraftVersion) {
+		StepWorker.Context context = new StepWorker.Context(repository, versionGraph, minecraftVersion);
+		StepWorker.Config config = new StepWorker.Config(GitCraft.config.getMappingsForMinecraftVersion(minecraftVersion).orElse(MappingFlavour.IDENTITY_UNMAPPED));
+
+		for (Step step : Step.values()) {
+			StepStatus status = null;
+			Exception exception = null;
+
+			long timeStart = System.nanoTime();
+
+			try {
+				MiscHelper.println("Performing step '%s' for %s (%s)...", step.getName(), context, config);
+				status = step.createWorker(config).run(this, context);
+			} catch (Exception e) {
+				status = StepStatus.FAILED;
+				exception = e;
+			}
+
+			long timeEnd = System.nanoTime();
+			long delta = timeEnd - timeStart;
+			Duration deltaDuration = Duration.ofNanos(delta);
+			String timeInfo = String.format("elapsed: %dm %02ds", deltaDuration.toMinutes(), deltaDuration.toSecondsPart());
+
+			switch (status) {
+			case SUCCESS ->
+				MiscHelper.println("\tStep '%s' for %s (%s) \u001B[32msucceeded\u001B[0m (%s)", step.getName(), context, config, timeInfo);
+			case UP_TO_DATE ->
+				MiscHelper.println("\tStep '%s' for %s (%s) was \u001B[32malready up-to-date\u001B[0m", step.getName(), context, config);
+			case NOT_RUN -> {
+				if (GitCraft.config.printNotRunSteps) {
+					MiscHelper.println("Step '%s' for %s (%s) was \u001B[36mnot run\u001B[0m", step.getName(), context, config);
+				}
+			}
+			}
+
+			if (status == StepStatus.FAILED) {
+				if (exception != null) {
+					MiscHelper.panicBecause(exception, "Step '%s' for %s (%s) \u001B[31mfailed\u001B[0m (%s)", step.getName(), context, config, timeInfo);
+				}
+			} else if (status.hasRun()) {
+				for (MinecraftJar minecraftJar : MinecraftJar.values()) {
+					StepResult resultFile = step.getMinecraftJar(minecraftJar);
+
+					// this step may not do anything with mc jars at all
+					if (resultFile != null) {
+						Path path = resultFiles.get(resultFile);
+
+						// this step may not have affected this mc jar
+						if (path != null) {
+							minecraftJars.put(minecraftJar, path);
+						}
+					}
+				}
```

Comment:
I'm not sure whether this replacing-approach of the minecraft jars after each step is sufficient.
Maybe a more explicit approach, like a `StepInput` (similar to a `StepResult`, would probably be an interface, each worker would have an input record implementing this, passed to the run function of a worker) could be used to reference artifacts from previous steps inside an executable pipeline (see other comment)?
This would also eliminate direct calls to results of other steps like in `ArtifactUnpacker` or `DataGenerator`, which would be nice.
Also, parallelization of steps would become harder when replacing minecraft jars after steps (although I don't see that happening anyway in the near future)

Reply:
> I'm not sure whether this replacing-approach of the minecraft jars after each step is sufficient.

I don't quite understand. Apart from parallelization, what is this hindering?

> This would also eliminate direct calls to results of other steps like in `ArtifactUnpacker` or `DataGenerator`, which would be nice.

I don't see this as a problem. In cases where the `StepResult`s are used directly, that's because a specific step's output is needed. Like the `ArtifactsUnpacker`, it specifically needs to get the unmodified artifacts. The `Decompiler` step specifically needs the libraries directory for that mc version, and so on.

Reply:
I'll revisit that (in detail) later, this is fine for now.
I thought that more flexibility might be needed in the future, in case multiple inputs are valid.